hack/e2e.sh

```diff
@@ -338,51 +339,85 @@ kubeadmConfigPatches:
     v: "4"
 EOF
     if [ -n "$DOCKER_IO_MIRROR" -o -n "$GCR_IO_MIRROR" -o -n "$QUAY_IO_MIRROR" ]; then
-cat <<EOF >> $tmpfile
+        cat <<EOF >>$tmpfile
 containerdConfigPatches:
 - |-
 EOF
         if [ -n "$DOCKER_IO_MIRROR" ]; then
-cat <<EOF >> $tmpfile
+            cat <<EOF >>$tmpfile
   [plugins."io.containerd.grpc.v1.cri".registry.mirrors."docker.io"]
     endpoint = ["$DOCKER_IO_MIRROR"]
 EOF
         fi
         if [ -n "$GCR_IO_MIRROR" ]; then
-cat <<EOF >> $tmpfile
+            cat <<EOF >>$tmpfile
   [plugins."io.containerd.grpc.v1.cri".registry.mirrors."gcr.io"]
     endpoint = ["$GCR_IO_MIRROR"]
 EOF
         fi
         if [ -n "$QUAY_IO_MIRROR" ]; then
-cat <<EOF >> $tmpfile
+            cat <<EOF >>$tmpfile
   [plugins."io.containerd.grpc.v1.cri".registry.mirrors."quay.io"]
     endpoint = ["$QUAY_IO_MIRROR"]
 EOF
         fi
     fi
     # control-plane
-    cat <<EOF >> $tmpfile
+    cat <<EOF >>$tmpfile
 nodes:
 - role: control-plane
 EOF
-    if [[ "$KIND_DATA_HOSTPATH" != "none" ]]; then
-        if [ ! -d "$KIND_DATA_HOSTPATH" ]; then
-            echo "error: '$KIND_DATA_HOSTPATH' is not a directory"
-            exit 1
+    # check $KIND_ETCD_DATADIR
+    if [[ "$KIND_ETCD_DATADIR" != "none" ]]; then
+        # check $KIND_DATA_HOSTPATH
+        if [[ "$KIND_DATA_HOSTPATH" != "none" ]]; then
+            if [ ! -d "$KIND_DATA_HOSTPATH" ]; then
+                echo "error: '$KIND_DATA_HOSTPATH' is not a directory"
+                exit 1
+            fi
+            if [ ! -d "$KIND_ETCD_DATADIR" ]; then
+                echo "error: '$KIND_ETCD_DATADIR' is not a directory"
+                exit 1
+            fi
+            local hostWorkerPath="${KIND_DATA_HOSTPATH}/control-plane"
+            test -d $hostWorkerPath || mkdir $hostWorkerPath
+            # set custom KIND_DATA_HOSTPATH and KIND_ETCD_DATADIR
+            cat <<EOF >>$tmpfile
+  extraMounts:
+  - containerPath: /mnt/disks/
+    hostPath: "$hostWorkerPath"
+    propagation: HostToContainer
+  - containerPath: /var/lib/etcd
+    hostPath: "$KIND_ETCD_DATADIR"
+EOF
+        else
+            # set custom KIND_ETCD_DATADIR
+            cat <<EOF >>$tmpfile
+  extraMounts:
+  - containerPath: /var/lib/etcd
+    hostPath: "$KIND_ETCD_DATADIR"
+EOF
         fi
-        local hostWorkerPath="${KIND_DATA_HOSTPATH}/control-plane"
-        test -d $hostWorkerPath || mkdir $hostWorkerPath
-        cat <<EOF >> $tmpfile
+    else
+        if [[ "$KIND_DATA_HOSTPATH" != "none" ]]; then
+            if [ ! -d "$KIND_DATA_HOSTPATH" ]; then
+                echo "error: '$KIND_DATA_HOSTPATH' is not a directory"
+                exit 1
+            fi
+            local hostWorkerPath="${KIND_DATA_HOSTPATH}/control-plane"
+            test -d $hostWorkerPath || mkdir $hostWorkerPath
+            # set custom KIND_DATA_HOSTPATH
+            cat <<EOF >>$tmpfile
```

Comment:
this part can be simplified like this:

1)  add `extraMounts:` if `[[ "$KIND_ETCD_DATADIR" != "none" || "$KIND_DATA_HOSTPATH" != "none"   ]]`
2) add mounts for `/mnt/disks/`  if KIND_DATA_HOSTPATH != "none"
3) add mounts for `/var/lib/etcd` if KIND_ETCD_DATADIR  != "none"

Reply:
btw, please revert format changes in other places, focus one thing in one PR. you can submit another PR to change the format.

Reply:
update